Add API for textual Code Actions

rubenporras opened the pull request:
In addition, improve naming of variables in the existing API to use
diagnostics for org.eclipse.lsp4j.Diagnostic instead of issues and
errors, and resolutions for DiagnosticResolution instead of issues in
the DiagnosticResolutionAcceptor

tivervac commented on org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/quickfix/DiagnosticResolution.java:
```diff
@@ -34,12 +41,22 @@ public class DiagnosticResolution {
 
 	private Options options;
 
+	public DiagnosticResolution(final String label, final Factory modificationContextFactory, final ITextModification<Diagnostic, EObject, Document, List<TextEdit>> textModification) {
```
No need for final in method parameters

tivervac replied:
Would also need an `@since` I suppose

tivervac replied:
You probably want a second constructor such that the `relevance` can be set

rubenporras replied:
done.

tivervac commented on org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/quickfix/ITextModification.java:
```diff
@@ -0,0 +1,19 @@
+ */
+@FunctionalInterface
+public interface ITextModification<A, B, C, R> {
+  R apply(A a, B b, C c);
```
Don't think any code uses two space as indent

tivervac replied:
What are a, b and c? Might be useful to give them better names or document them

szarnekow replied:
Also I don't feel like the current ITextModification looks like an ITextModification at all. More like a Function3.

Judging from the usages, the parameterizations are the same for all. Maybe remove the unnecessary type parameters.

rubenporras replied:
Done, i have removed the type parameters

tivervac commented on org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/TextEditAcceptor.java:
```diff
@@ -0,0 +1,76 @@
+	private final boolean useDocumentChanges;
+	private final WorkspaceEdit edit;
+
+	public TextEditAcceptor(final WorkspaceEdit edit, final ILanguageServerAccess access) {
```
No need for final parameters

rubenporras replied:
done, we have different policy according to tabs and spaces and usages of final at Avaloq, and these slipped through when coding.

rubenporras commented:
@cdietrich, @tivervac, from my point of view I am like the change as it is now. I have cleaned up the warnings and added one test. 

Would you mind taking a look? 

I would be happy if we could merge it before M2 is out so that I can upgrade and use the new API.

Thanks

szarnekow commented on org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/quickfix/DiagnosticResolutionAcceptor.java:
```diff
@@ -35,12 +35,16 @@ public DiagnosticResolutionAcceptor(DiagnosticModificationContext.Factory modifi
+		resolutions.add(new DiagnosticResolution(label, modificationContextFactory, modification));
+	}
+
+	public void accept(final String label, final ITextModification modification) {
```
Nit: final parameters are not adding value here. Not actionable.

You may want to use Oomph to setup your Xtext dev environment, to simplify the review process for you and the reviewer though :)

rubenporras replied:
Sorry for that. I did that this morning already to write the test, it was just too error-prone to do that in GitHub editor :) It will be part of the next push.

szarnekow commented on org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/quickfix/IQuickFixProvider.java:
```diff
@@ -33,10 +33,10 @@ public interface IQuickFixProvider {
-	 * @param issue
-	 *            EMF diagnostic issue
+	 * @param diagnostic
+	 *            the diagnostic
```
Thank you for the attention to detail.

szarnekow commented on org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/quickfix/ITextModification.java:
```diff
@@ -0,0 +1,26 @@
+ */
+@FunctionalInterface
+public interface ITextModification {
+	List<TextEdit> apply(Diagnostic diagnostic, EObject object, Document document);
```
Please add a comment that describes which `EObject object` one would expect here.

szarnekow commented on org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/quickfix/DiagnosticResolution.java:
```diff
@@ -61,28 +85,31 @@ public String getLabel() {
+				serializer.applyModifications(changeConverter);
+			} else {
+				TextEditAcceptor textEditAcceptor = new TextEditAcceptor(edit, access);
+				Diagnostic diagnostic = params.getContext().getDiagnostics().get(0);
```
I'm a little scared about the (`.get(0)`) here. This is probably incorrect.

rubenporras replied:
The code in [org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/codeActions/QuickFixCodeActionService.java](https://github.com/eclipse/xtext-core/blob/942def122c1b49feac1713de2edeae514b0b1b7c/org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/codeActions/QuickFixCodeActionService.java#L69-71) makes sure that we have a singleton list, with exactly one element, so as the code currently is, I do not think this is a mistake, as it relies on an invariant, but I also dislike it. I left the code as it is because I am not sure what should be better than this.
Should we add a precondition check that tests that the list has indeed one and only one argument? Would you prefer that?

szarnekow replied:
Yes, please use something like com.google.common.collect.Iterables.getOnlyElement(Iterable<T>) to document and assert that invariant in code.

rubenporras replied:
Nice, I was not aware of that Guava API. I have opted for using what they document as the variant for Java 8 using streams.

szarnekow commented on org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/TextEditAcceptor.java:
```diff
@@ -0,0 +1,74 @@
+		}
+	}
+
+	public Object accept(String theUri, Document document, List<TextEdit> textEdits) {
```
This should be a void method as most of the Acceptor-style APIs in Xtext are?

Currently it either return `true` or a random old textEdits or null, none of these do make sense to me.

rubenporras replied:
done

szarnekow commented on org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/rename/ChangeConverter2.java:
```diff
@@ -163,17 +130,7 @@ protected void _handleReplacements(ITextDocumentChange change) {
 	}
 
 	protected Object addTextEdit(String theUri, Document document, TextEdit... textEdits) {
```
Please refactor to void method, too. Probably fallout from some Xtend to Java rewrite in the past. line 124 can always return null.

cdietrich replied:
i assume the problem is that the doRead calls want a return value. what do you propose?

szarnekow replied:
I'd let this particular doRead either return null or make it an IUnitOfWork.Void

cdietrich replied:
i guess an explict return null would do the latter

rubenporras replied:
this would be a breaking change if someone relies on it, but if it is fine for you, it is for me as well, so I have done the change. Thanks

szarnekow reviewed (commented):
Thank you Ruben. Please let me know if you disagree with some of the comments.

Do you see - by chance - a path towards having a common APi for Eclipse Quickfixes and LSP Quickfixes based on the API that is introduced here?

cdietrich commented on org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/rename/ChangeConverter2.java:
```diff
@@ -120,7 +87,8 @@ protected void _handleReplacements(IEmfResourceChange change) {
 				TextEdit textEdit = new TextEdit(range, newContent);
-				return addTextEdit(uri, document, textEdit);
+				addTextEdit(uri, document, textEdit);
+				return null;
```
:+1:

tivervac commented on org.eclipse.xtext.ide.tests/src/org/eclipse/xtext/ide/tests/editor/quickfix/AbstractIdeQuickfixTest.java:
```diff
@@ -200,14 +200,14 @@ public ResourceSet newLiveScopeResourceSet(URI uri) {
+					quickFixProvider.getResolutions(options, issue).stream()
+					.filter(r -> r.getLabel().equals(expectedIssueResolution.getLabel()))
+					.collect(Collectors.toList());
+			assertEquals("More than one quickfixes available!", 1, actualIssueResolutions.size());
```
s/quickfixes/quickfix/

rubenporras replied:
done

tivervac commented on org.eclipse.xtext.ide.tests/testlang-src/org/eclipse/xtext/ide/tests/testlanguage/ide/quickfix/TestLanguageQuickFixProvider.java:
```diff
@@ -17,12 +20,26 @@
-			final TypeDeclaration element = (TypeDeclaration) obj;
-			element.setName(StringExtensions.toFirstUpper(element.getName()));
+		acceptor.accept(EMF_QF_LABEL, obj -> {
+			final TypeDeclaration declaration = (TypeDeclaration) obj;
```
Probably also from x2j, no need for the final

rubenporras replied:
done

tivervac commented on org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/quickfix/ITextModification.java:
```diff
@@ -0,0 +1,39 @@
+	 * @param diagnostic
+	 *            the {@link Diagnostic}
+	 * @param object
+	 *            the {@link EObject} next to the offset at the start of the range of the diagnostic
```
This is a bit hard to read, how about 

`the {@link EObject} after {@code diagnostic.getRange().getStart()}`

tivervac commented on org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/quickfix/ITextModification.java:
```diff
@@ -0,0 +1,39 @@
+public interface ITextModification {
+	/**
+	 *
+	 * Returns a list of {@link TextEdit} given a diagnostic, object and document.
```
Missing `s` after `{@link TextEdit}`

rubenporras replied:
done

tivervac commented on org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/codeActions/QuickFixCodeActionService.java:
```diff
@@ -44,26 +46,28 @@ public class QuickFixCodeActionService implements ICodeActionService2 {
+		List<Either<Command, CodeAction>> result = new ArrayList<>();
+		for (Diagnostic diagnostic : options.getCodeActionParams().getContext().getDiagnostics()) {
+			List<DiagnosticResolution> resolutions = quickfixes.getResolutions(options, diagnostic).stream()
+					.sorted(Comparator.comparing(DiagnosticResolution::getLabel)).collect(Collectors.toList());
```
Not sure this comparator will work. I had a quick look at the uses of `getLabel` and I see some places null checking it. The datastructure `DiagnosticResolution` does not make sure that the label is not null and as far as I can see, `Comparator.comparing` can't handle keyExtractors that can produce nulls

tivervac commented on org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/rename/ChangeConverter2.java:
```diff
@@ -120,7 +87,8 @@ protected void _handleReplacements(IEmfResourceChange change) {
+				return null;
 			}).get();
 		} catch (InterruptedException | ExecutionException | IOException e) {
 			throw Exceptions.sneakyThrow(e);
```
Not your code, but `sneakyThrow` is evil. It will disguise checked Exceptions as unchecked exceptions and throw them without mentioning the checked exceptions in a function's signature.

Since you have to edit the signature to correctly handle this, not sure if there's a good solution here

rubenporras replied:
Thanks. I would be happy changing it, but since it is unrelated to my changes, I would leave it be unless someone wants me to do the change.

tivervac reviewed (commented):
LGTM now

cdietrich commented:
@szarnekow ok to go?

szarnekow reviewed (approved):
Thank you @rubenporras

szarnekow commented:
> @szarnekow , at the moment I do not see a path towards having a common API [..]

It does make sense, yes. Thank you for sharing your thoughts on this.

cdietrich commented:
thx @rubenporras
